Rework whitehall-frontend breadcrumb to use content store

mobaig opened the pull request:
Use content store instead of content API. Core logic that needs reworking is in SpecialistTagFinder.

This is intended to lead up to work to remove specialist_sector related DB fields in the Whitehall DB, and to migrate the source of truth for all documents to being the publishing API.

https://trello.com/c/f6coMIbz

boffbowsh commented on lib/specialist_tag_finder.rb:
```diff
@@ -1,32 +1,31 @@
-    if primary_tag_slug = @document.primary_specialist_sector_tag
-      specialist_sector_tags.find {|t| t.slug == primary_tag_slug }
-    end
+  def grandparent_topic
```
Do you happen to know why it's the grandparent topic? Is it always the grandparent? This is an open question here: https://github.com/alphagov/content-store/pull/203

mobaig replied:
Looks like this has been covered by @tvararu in #203.

boffbowsh commented on lib/specialist_tag_finder.rb:
```diff
@@ -1,32 +1,31 @@
+    parents = Array(edition_content_item.links["parent"])
+    return unless parents.any?
+
+    # FIXME: We now need to fetch the parent topic from the content store to
```
You may be able to retrieve them both at once from content store, making it do the resolution. See https://github.com/alphagov/content-store/pull/203

Davidslv commented on app/helpers/document_helper.rb:
```diff
@@ -217,8 +217,11 @@ def part_of_metadata(document, policies = [], sector_tag_finder = nil)
+      links_to_topics = tagged_topics.map do |topic|
+        link_to topic.title, topic.web_url, class: 'sector-link'
+      end
+      part_of += links_to_topics
```
Wouldn't it be better to use `<<` instead of `+=` ?

http://ruby-doc.org/core-1.9.3/Array.html#method-i-2B

> Concatenation—Returns a new array built by concatenating the two arrays together to produce a third array.

http://ruby-doc.org/core-1.9.3/Array.html#method-i-3C-3C

> Append—Pushes the given object on to the end of this array. This expression returns the array itself, so several appends may be chained together.

tijmenb replied:
That would leave you with a nested array.

```ruby
[1] + [2] = [1, 2] 
[1] << [2] = [1, [2]]
```

Davidslv replied:
Ah! I should have tested it before.

MatMoore commented on app/helpers/document_helper.rb:
```diff
@@ -217,8 +217,11 @@ def part_of_metadata(document, policies = [], sector_tag_finder = nil)
 
-    if sector_tag_finder && (all_sectors = sector_tag_finder.sectors_and_subsectors).any?
-      part_of += array_of_links_to_sectors(all_sectors)
+    if sector_tag_finder && (tagged_topics = sector_tag_finder.topics).any?
```
Nitpick: I know this was there before, but as a general style thing I'm not a fan of assignment in the middle of expressions. It's a really common mistake to type `=` when you mean `==`, so whenever I read this I do a double take and have to check whether its actually intentional.

Python [explicitly made this a syntax error](https://docs.python.org/2/faq/design.html#why-can-t-i-use-an-assignment-in-an-expression), and having just spent ~30 minutes debugging a similar problem, I really miss this feature in ruby!

mobaig replied:
Agreed, not a fan either. Will rework.

MatMoore commented on lib/specialist_tag_finder.rb:
```diff
@@ -1,32 +1,33 @@
-
-  def artefact
-    @artefact ||= Whitehall.content_api.artefact(RegisterableEdition.new(@document).slug)
+  def grandparent_topic
```
It might be good to add a comment explaining what the grandparent topic is, or perhaps rename it to clarify that it is the top level topic the edition falls under. I'm not sure I understand it correctly. I'm assuming:

* The parent link is always a topic
* The topic hierarchy is 2 levels deep
* We can only tag to the bottom level
* We only display the top level in the breadcrumb (previously known as the primary specialist sector)

E.g. an edition parent would be environmental-management/chemicals and then the grandparent is environmental-management

Is this correct?

mobaig replied:
Noted - will add an explanatory comment.

mobaig replied:
Oh - that's correct btw.

MatMoore commented:
Added a few minor comments inline but otherwise this looks good 👍 

Based on my current understanding of the architecture, I think we should ultimately be getting these links from the publishing api rather than content store, but I'm assuming this is blocked on links migration and/or dependency resolution work?

mobaig commented:
Thanks @MatMoore , addressing your comments now. Re: your question about source of the links - given this is `whitehall-frontend`-related logic, I think it's architecturally consistent for a frontend app to be getting its data from the content store. 

I'm not sure if there's any intention to change this behaviour so that frontends communicate with the publishing API instead.

Irrespective of the source, the dependency resolution work is a precursor for making the code here suck less.

MatMoore commented:
> given this is whitehall-frontend-related logic, I think it's architecturally consistent for a frontend app to be getting its data from the content store.

Ah yeah, of course it is. For some reason I was thinking this was backend thing. 🐌

mobaig commented:
@MatMoore @tijmenb @Davidslv (and anyone else for that matter) - have addressed comments in a series of commits from https://github.com/alphagov/whitehall/pull/2557/commits/b7130656d9daef96c3f0dd4dfd42247c09825e23 onwards. What do you reckon?

Davidslv commented on features/support/specialist_sector_helper.rb:
```diff
@@ -1,7 +1,19 @@
 require 'gds_api/test_helpers/content_api'
+require "gds_api/test_helpers/content_store"
+
+Before do
```
I honestly didn't know a capitalize `before` would work, even though (sorry I know it's nitpicking) could you change it to lowercase please ?

mobaig replied:
That's how you write before hooks in Cucumber. https://www.relishapp.com/cucumber/cucumber/docs/writing-support-code/before-hook

Davidslv replied:
Oh. Ignore my comment, I thought I was reading an rspec spec ... **confused**

Davidslv commented on app/views/documents/_header.html.erb:
```diff
@@ -7,10 +7,16 @@
-                    heading: document.title,
-                    extra: true } %>
+<%=
+	render(
```
could we remove the extra tab please?
It looks good and clean actually, but I'm sure it's not consistent with the rest of the codebase in Whitehall

mobaig replied:
Ugh, half-baked vim config. Updated.

tijmenb commented on features/step_definitions/specialist_sector_steps.rb:
```diff
@@ -13,15 +13,34 @@
+      "topics" => [{ "title" => "Topic 1" }, { "title" => "Topic 2"  }]
+    }
+  })
+  parent_content_item = content_item_for_base_path(parent_base_path).merge!({
```
Is there a reason for the mutating `merge!` here?

tijmenb commented:
Other than my question about `merge!` vs `merge`, this looks good! Since we are going to run a before/after test using [govuk-diff-pages](https://github.com/alphagov/govuk-diff-pages) on staging, this can be merged.

👍
